src/livestatus.c

```diff
@@ -0,0 +1,468 @@
+      .neb_callbacks = strtoull(fields[13], NULL, 10),
+      .neb_callbacks_rate = strtod(fields[14], NULL),
+      .requests = strtoull(fields[15], NULL, 10),
+      .requests_rate = strtod(fields[16], NULL),
+      .service_checks = strtoull(fields[17], NULL, 10),
+      .service_checks_rate = strtod(fields[18], NULL),
+  };
+
+  return status;
+} /* livestatus_status_t c_to_livestatus_status */
+
+// Borrowed from https://stackoverflow.com/a/29378380
+static long ls_strtol_subrange(const char *s, char **endptr, int base, long min,
+                               long max) {
+  long y = strtol(s, endptr, base);
+  if (y > max) {
+    errno = ERANGE;
+    return max;
+  }
+  if (y < min) {
+    errno = ERANGE;
+    return min;
+  }
+  return y;
+} /* static long ls_strtol_subrange */
+
+// Borrowed from https://stackoverflow.com/a/29378380
+static int ls_strtoi(const char *s, char **endptr, int base) {
+#if INT_MAX == LONG_MAX && INT_MIN == LONG_MIN
+  return (int)strtol(s, endptr, base);
+#else
+  return (int)ls_strtol_subrange(s, endptr, base, INT_MIN, INT_MAX);
+#endif
+} /* static int ls_strtoi */
+
+static int ls_config(const char *key, const char *value) /* {{{ */
+{
+  if (strcasecmp("LivestatusSocketFile", key) == 0) {
+    char *socket_file = strdup(value);
+    if (socket_file == NULL) {
+      ERROR("livestatus plugin: strdup failed: %s", STRERRNO);
+      return -1;
+    }
+
+    livestatus_obj.socket_file = socket_file;
+  }
+
+  if (strcasecmp("OnFailureMaxRetry", key) == 0) {
+    livestatus_obj.max_retry = ls_strtoi(value, NULL, 10);
+    if (errno != 0) {
+      ERROR("livestatus plugin: strtoi failed for OnFailureMaxRetry: %s",
+            STRERRNO);
+      return -1;
+    }
+  }
+
+  if (strcasecmp("OnFailureBackOffSeconds", key) == 0) {
+    livestatus_obj.backoff_sec = ls_strtoi(value, NULL, 10);
+    if (errno != 0) {
+      ERROR("livestatus plugin: strtoi failed for OnFailureBackOffSeconds: %s",
+            STRERRNO);
+      return -1;
+    }
+  }
+
+  return 0;
+} /* static int ls_config */
+
+static int unix_connect(const char *sockfile, int *sockfd) {
+  struct sockaddr_un sun = {
+      .sun_family = AF_UNIX,
+  };
+  int rc = -1;
+  int sfd = -1;
+
+  sfd = socket(AF_UNIX, SOCK_STREAM, 0);
+  if (sfd < 0) {
+    ERROR("livestatus plugin: creating client socket: %s", STRERRNO);
+    return -1;
+  }
+
+  sstrncpy(sun.sun_path, sockfile, strlen(sockfile));
+
+  do {
+    rc = connect(sfd, (struct sockaddr *)&sun, sizeof(sun));
+  } while (errno == EINTR);
+
+  if (rc < 0) {
+    ERROR("livestatus plugin: connect to socket file %s: %s", sockfile,
+          STRERRNO);
+    return -1;
+  }
+
+  *sockfd = sfd;
+
+  return 0;
+} /* unix_connect */
+
+static int ls_parse(const char *lresponse, livestatus_status_t *lstatus) {
+  int rc = -1;
+  int i = 0;
+  char **fields = NULL;
+  char *pchar = NULL;
+
+  fields = malloc(sizeof(char **) * LIVESTATUS_EXPECTED_FIELDS_RESP_NB);
```

Comment:
I the readability of this function can be greatly improved using `strarray_add` and `strarray_free` from `"common.h"`:

```c
char **fields = NULL;
size_t fields_num = 0;
// …
while (read fields) {
  char *value = strtok(); // see below
  strarray_add(&fields, &fields_num, value);
}
// …
strarray_free(fields, fields_num);
return rc;
```

Reply:
*Obsolete:* You pushed new commits while I was reviewing. I've copied this comment and created a new comment attached to the new location.